### ISSUE TITLE
Expose ability to set blob metadata in GCSHook

### ISSUE THREAD

**Issue by ejstembler** [apache/airflow#16487]
**Description**

Expose ability to set blob metadata in the [GCSHook](https://github.com/apache/airflow/blob/main/airflow/providers/google/cloud/hooks/gcs.py). Probably via the [upload](https://github.com/apache/airflow/blob/main/airflow/providers/google/cloud/hooks/gcs.py#L393) method with a new parameter:  `metadata: Optional[Dict[str, str]] = None`.

**Use case / motivation**

As a best practice, I always set blob metadata attributes which are useful or provide more information.  Any information about the blob, such as tracking information, foreign keys, job information, etc.

Using the Google SDK, this is what it looks like:

```python
metadata = {
    "job-id": job_id,
    "user-id": user_id,
    "batch-size": batch_size,
    # ...
}

storage_client = storage.Client()
bucket = storage_client.bucket(gcp_bucket_name)
blob = bucket.blob(destination_blob_name)
blob.content_type = content_type
blob.metadata = metadata
```

**Comment by potiuk** [comment]
Would you be willing to make a PR about this ? Sounds like nice first contribution

**Comment by ejstembler** [comment]
@potiuk I think it's a small change. i.e. a new parameter and 1 additional line of code:

```python
if metadata:
    blob.blob.metadata = metadata
```

Plus changes to any related unit tests.

What's the repo/contribution policy on changing function arity though?  Will this create breaking changes?

**Comment by potiuk** [comment]
Airflow is community driven, so it is normal for users like you to contribute. Many do. It's a nice way to give back to the community. 

See CONTRIBITING.rst in the root of the repo (the repo is exactly the one we are discussing it).

**Comment by potiuk** [comment]
And i believe you can make it in non-breaking way. Airflow requires all the operator init args.to be keyword args, so adding a new keyword with default should not be breaking

**Comment by eladkal** [comment]
@ejstembler would you like to finish the PR you started?

**Comment by ejstembler** [comment]
Hi @eladkal. I'm not sure what is left.  In the PR [comments](https://github.com/apache/airflow/pull/16492#issuecomment-863196996), @uranusjr asked for a test. However, I couldn't find an adequate way to test since the relative object was internal to a function and transient. Ref: https://github.com/apache/airflow/pull/16492#issuecomment-863781549

**Comment by shahar1** [comment]
This issue was recreated in #21978 and resolved by #22058

**Comment by eladkal** [comment]
Thanks @shahar1
